flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java

```diff
@@ -814,6 +820,69 @@ public Executor getFutureExecutor() {
 				entry -> serializeAccumulator(entry.getKey(), entry.getValue())));
 	}
 
+	@Override
+	public BlockingPersistentResultPartitionMeta getBlockingPersistentResultPartitionMeta() {
+		Map<AbstractID, Map<AbstractID, ResultPartitionDescriptor>> resultPartitionDescriptors = new HashMap<>();
+
+		// keep record of all failed IntermediateDataSetID
+		Set<AbstractID> failedIntermediateDataSetIds = new HashSet<>();
+
+		for (ExecutionVertex executionVertex : getAllExecutionVertices()) {
+			for (IntermediateResultPartition intermediateResultPartition : executionVertex.getProducedPartitions().values()) {
+				if (intermediateResultPartition.getResultType() == ResultPartitionType.BLOCKING_PERSISTENT) {
+					try {
+						addLocation(resultPartitionDescriptors, intermediateResultPartition);
+					} catch (Throwable throwable) {
+						LOG.error("Failed to get location of ResultPartition: " + intermediateResultPartition.getPartitionId(), throwable);
+						failedIntermediateDataSetIds.add(
+							new AbstractID(intermediateResultPartition.getIntermediateResult().getId()));
+					}
+				}
+			}
+		}
+
+		return new BlockingPersistentResultPartitionMeta(resultPartitionDescriptors, failedIntermediateDataSetIds);
+	}
+
+	/**
+	 *
+	 * @param resultPartitionDescriptors
+	 * @param intermediateResultPartition
+	 * throw exception if any error occurs.
+	 */
+	public void addLocation(
```

Comment:
This utility method should not be in the ExecutionGraph - it just blows up the public signature.
It should be private static or moved to a utility class.